File: services/src/manage/network.rs
// Copyright (C) 2023 Huawei Device Co., Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
use std::sync::{Arc, RwLock};

use cxx::UniquePtr;
use ffi::NetworkRegistry;
pub(crate) use ffi::{NetworkInfo, NetworkType};
use NetworkState::{Offline, Online};

use crate::manage::network_manager::NetworkManager;

cfg_oh! {
    use super::events::TaskManagerEvent;
    use super::task_manager::TaskManagerTx;
}

#[derive(Clone)]
pub struct Network {
    pub(crate) inner: NetworkInner,
    #[cfg(feature = "oh")]
    pub(crate) _registry: Option<Arc<UniquePtr<NetworkRegistry>>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum NetworkState {
    Offline,
    Online(NetworkInfo),
}

impl Network {
    pub(crate) fn state(&self) -> NetworkState {
        self.inner.state.read().unwrap().clone()
    }
}

pub(crate) fn register_network_change() {
    const RETRY_TIME: i32 = if cfg!(test) { 1 } else { 20 };
    let mut count: i32 = 0;
    let mut network_manager = NetworkManager::get_instance().lock().unwrap();
    let tx = network_manager.tx.clone();
    if network_manager.network.state() != Offline {
        return;
    }
    match tx {
        Some(tx) => {
            let mut registry: UniquePtr<NetworkRegistry> = UniquePtr::null();
            while count < RETRY_TIME {
                registry = ffi::RegisterNetworkChange(
                    Box::new(network_manager.network.inner.clone()),
                    Box::new(NetworkTaskManagerTx { inner: tx.clone() }),
                    |task_manager| {
                        task_manager.inner.send_event(TaskManagerEvent::network());
                    },
                    |task_manager| {
                        task_manager.inner.send_event(TaskManagerEvent::network());
                    },
                );
                if registry.is_null() {
                    std::thread::sleep(std::time::Duration::from_secs(1));
                    count += 1;
                    continue;
                }
                break;
            }
            if registry.is_null() {
                error!("RegisterNetworkChange failed!");
                return;
            }
            network_manager.network._registry = Some(Arc::new(registry));
        }
        None => {
            error!("register_network_change failed, tx is None!");
        }
    }
}

#[derive(Clone)]
pub struct NetworkInner {
    state: Arc<RwLock<NetworkState>>,
}

pub struct NetworkTaskManagerTx {
    #[cfg(feature = "oh")]
    inner: TaskManagerTx,
}

impl NetworkInner {
    pub(crate) fn new() -> Self {
        Self {
            state: Arc::new(RwLock::new(NetworkState::Offline)),
        }
    }

    pub(crate) fn notify_offline(&self) {
        let mut state = self.state.write().unwrap();
        if *state != Offline {
            info!("network is offline");
            *state = Offline;
        }
    }

    pub(crate) fn notify_online(&self, info: NetworkInfo) -> bool {
        let mut state = self.state.write().unwrap();
        if !matches!(&*state, Online(old_info) if old_info == &info  ) {
            info!("network online {:?}", info);
            *state = Online(info.clone());
            true
        } else {
            false
        }
    }
}

unsafe impl Send for NetworkRegistry {}
unsafe impl Sync for NetworkRegistry {}

#[allow(unreachable_pub)]
#[cxx::bridge(namespace = "OHOS::Request")]
mod ffi {

    #[derive(Clone, Eq, PartialEq, Debug)]
    struct NetworkInfo {
        network_type: NetworkType,
        is_metered: bool,
        is_roaming: bool,
    }

    #[repr(u8)]
    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
    enum NetworkType {
        Other,
        Wifi,
        Cellular,
    }

    extern "Rust" {
        type NetworkInner;
        type NetworkTaskManagerTx;
        fn notify_online(self: &NetworkInner, info: NetworkInfo) -> bool;
        fn notify_offline(self: &NetworkInner);
    }

    unsafe extern "C++" {
        include!("network.h");
        include!("c_request_database.h");
        type NetworkRegistry;
        fn RegisterNetworkChange(
            notifier: Box<NetworkInner>,
            task_manager: Box<NetworkTaskManagerTx>,
            notify_online: fn(&NetworkTaskManagerTx),
            notify_offline: fn(&NetworkTaskManagerTx),
        ) -> UniquePtr<NetworkRegistry>;
    }
}

#[cfg(feature = "oh")]
#[cfg(test)]
mod test {
    use super::*;
    use crate::tests::test_init;

    #[test]
    fn ut_network() {
        test_init();
        let notifier;
        {
            let network_manager = NetworkManager::get_instance().lock().unwrap();
            notifier = network_manager.network.inner.clone();
        }
        assert!(!NetworkManager::is_online());

        notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Wifi,
            is_metered: false,
            is_roaming: false,
        });
        assert!(NetworkManager::is_online());
        assert_eq!(
            NetworkManager::query_network(),
            Online(NetworkInfo {
                network_type: NetworkType::Wifi,
                is_metered: false,
                is_roaming: false,
            })
        );
        notifier.notify_offline();
        assert!(!NetworkManager::is_online());
        notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Cellular,
            is_metered: true,
            is_roaming: true,
        });
        assert!(NetworkManager::is_online());
        assert_eq!(
            NetworkManager::query_network(),
            Online(NetworkInfo {
                network_type: NetworkType::Cellular,
                is_metered: true,
                is_roaming: true,
            })
        );
    }

    #[test]
    fn ut_network_notify() {
        test_init();
        let notifier = NetworkInner::new();
        notifier.notify_offline();
        assert!(notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Wifi,
            is_metered: true,
            is_roaming: true,
        }));
        assert!(!notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Wifi,
            is_metered: true,
            is_roaming: true,
        }));
        assert!(notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Wifi,
            is_metered: false,
            is_roaming: true,
        }));
        assert!(notifier.notify_online(NetworkInfo {
            network_type: NetworkType::Cellular,
            is_metered: false,
            is_roaming: true,
        }));
    }
}

File: services/src/cxx/network.cpp
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "network.h"

#include <cstdint>

#include "cxx.h"
#include "log.h"
#include "manage/network.rs.h"
#include "net_all_capabilities.h"
#include "net_conn_callback_stub.h"
#include "net_conn_client.h"
#include "net_specifier.h"
#include "refbase.h"

#ifdef REQUEST_TELEPHONY_CORE_SERVICE
#include "cellular_data_client.h"
#include "core_service_client.h"
#endif

#ifdef REQUEST_TELEPHONY_CORE_SERVICE
#include "iservice_registry.h"
#include "network_state.h"
#include "system_ability_definition.h"
#include "telephony_errors.h"
#endif

namespace OHOS::Request {
using namespace OHOS::NetManagerStandard;

RequestNetCallbackStub::RequestNetCallbackStub(
    rust::box<NetworkInner> network, rust::box<NetworkTaskManagerTx> task_manager,
    rust::fn<void(const NetworkTaskManagerTx &task_manager)> notifyTaskManagerOnline,
    rust::fn<void(const NetworkTaskManagerTx &task_manager)> notifyTaskManagerOffline

)
{
    networkNotifier_ = network.into_raw();
    task_manager_ = task_manager.into_raw();
    notifyTaskManagerOnline_ = notifyTaskManagerOnline;
    notifyTaskManagerOffline_ = notifyTaskManagerOffline;
}

RequestNetCallbackStub::~RequestNetCallbackStub()
{
    rust::Box<NetworkInner>::from_raw(networkNotifier_);
    rust::Box<NetworkTaskManagerTx>::from_raw(task_manager_);
}

void RequestNetCallbackStub::HandleNetCap(const sptr<NetAllCapabilities> &netAllCap)
{
    if (netAllCap->netCaps_.find(NetCap::NET_CAPABILITY_VALIDATED) == netAllCap->netCaps_.end()) {
        networkNotifier_->notify_offline();
        notifyTaskManagerOffline_(*task_manager_);
        return;
    }

    for (auto bearerType : netAllCap->bearerTypes_) {
        auto networkInfo = NetworkInfo();
        if (bearerType == NetManagerStandard::NetBearType::BEARER_WIFI) {
            networkInfo.network_type = NetworkType::Wifi;
            networkInfo.is_metered = false;
            networkInfo.is_roaming = false;

            if (networkNotifier_->notify_online(networkInfo)) {
                notifyTaskManagerOnline_(*task_manager_);
            }
            return;
        } else if (bearerType == NetManagerStandard::NetBearType::BEARER_CELLULAR) {
            networkInfo.network_type = NetworkType::Cellular;
            networkInfo.is_metered = true;
            networkInfo.is_roaming = this->IsRoaming();

            if (networkNotifier_->notify_online(networkInfo)) {
                notifyTaskManagerOnline_(*task_manager_);
            }
            return;
        };
    }
    if (networkNotifier_->notify_online(NetworkInfo{
            .network_type = NetworkType::Other,
            .is_metered = false,
            .is_roaming = false,
        })) {
        notifyTaskManagerOnline_(*task_manager_);
    }
    return;
}

int32_t RequestNetCallbackStub::NetAvailable(sptr<NetHandle> &netHandle)
{
    sptr<NetAllCapabilities> netAllCap = sptr<NetAllCapabilities>::MakeSptr();
    int32_t ret = NetConnClient::GetInstance().GetNetCapabilities(*netHandle, *netAllCap);
    if (ret != 0) {
        REQUEST_HILOGE("GetNetCapabilities failed, ret = %{public}d", ret);
        return ret;
    }
    this->HandleNetCap(netAllCap);
    return 0;
}

int32_t RequestNetCallbackStub::NetLost(sptr<NetHandle> &netHandle)
{
    networkNotifier_->notify_offline();
    notifyTaskManagerOffline_(*task_manager_);
    return 0;
}

int32_t RequestNetCallbackStub::NetUnavailable()
{
    networkNotifier_->notify_offline();
    notifyTaskManagerOffline_(*task_manager_);
    return 0;
}

int32_t RequestNetCallbackStub::NetCapabilitiesChange(
    sptr<NetHandle> &netHandle, const sptr<NetAllCapabilities> &netAllCap)
{
    REQUEST_HILOGI("NetCapabilitiesChange");
    this->HandleNetCap(netAllCap);
    return 0;
}

bool RequestNetCallbackStub::IsRoaming()
{
#ifdef REQUEST_TELEPHONY_CORE_SERVICE
    REQUEST_HILOGD("upload roaming");
    // Check telephony SA.
    {
        std::lock_guard<std::mutex> lock(roamingMutex_);

        auto sm = SystemAbilityManagerClient::GetInstance().GetSystemAbilityManager();
        if (sm == nullptr) {
            REQUEST_HILOGE("GetSystemAbilityManager return null");
            return false;
        }
        auto systemAbility = sm->CheckSystemAbility(TELEPHONY_CORE_SERVICE_SYS_ABILITY_ID);
        if (systemAbility == nullptr) {
            REQUEST_HILOGE("Telephony SA not found");
            return false;
        }
    }

    constexpr int32_t INVALID_SLOT_ID = -1;
    int32_t maxSlotNum = DelayedRefSingleton<OHOS::Telephony::CoreServiceClient>::GetInstance().GetMaxSimCount();
    bool isSim = false;
    for (int32_t i = 0; i < maxSlotNum; ++i) {
        if (DelayedRefSingleton<OHOS::Telephony::CoreServiceClient>::GetInstance().IsSimActive(i)) {
            isSim = true;
            break;
        }
    }
    if (!isSim) {
        REQUEST_HILOGD("no sim");
        return false;
    }

    int32_t slotId =
        DelayedRefSingleton<OHOS::Telephony::CellularDataClient>::GetInstance().GetDefaultCellularDataSlotId();
    if (slotId <= INVALID_SLOT_ID) {
        REQUEST_HILOGE("GetDefaultCellularDataSlotId InValidData");
        return false;
    }
    sptr<OHOS::Telephony::NetworkState> networkClient = nullptr;
    DelayedRefSingleton<OHOS::Telephony::CoreServiceClient>::GetInstance().GetNetworkState(slotId, networkClient);
    if (networkClient == nullptr) {
        REQUEST_HILOGE("networkState is nullptr");
        return false;
    }
    REQUEST_HILOGI("Roaming = %{public}d", networkClient->IsRoaming());
    return networkClient->IsRoaming();
#else
    REQUEST_HILOGE("Telephony SA not found");
    return false;
#endif
}

std::unique_ptr<NetworkRegistry> RegisterNetworkChange(rust::box<NetworkInner> notifier,
    rust::box<NetworkTaskManagerTx> task_manager,
    rust::fn<void(const NetworkTaskManagerTx &task_manager)> notifyTaskManagerOnline,
    rust::fn<void(const NetworkTaskManagerTx &task_manager)> notifyTaskManagerOffline)
{
    REQUEST_HILOGI("RegisterNetworkChange");
    sptr<RequestNetCallbackStub> callbackStub = sptr<RequestNetCallbackStub>::MakeSptr(
        std::move(notifier), std::move(task_manager), notifyTaskManagerOnline, notifyTaskManagerOffline);
    if (callbackStub == nullptr) {
        REQUEST_HILOGE("callbackStub is nullptr");
        return nullptr;
    }
    int ret = NetConnClient::GetInstance().RegisterNetConnCallback(callbackStub);
    if (ret != 0) {
        REQUEST_HILOGE("RegisterNetConnCallback failed, ret = %{public}d", ret);
        return nullptr;
    }
    return std::make_unique<NetworkRegistry>(callbackStub);
}

NetworkRegistry::NetworkRegistry(sptr<RequestNetCallbackStub> callback)
{
    callback_ = callback;
}

NetworkRegistry::~NetworkRegistry()
{
    REQUEST_HILOGI("UnregisterNetworkChange");
    int32_t ret = NetConnClient::GetInstance().UnregisterNetConnCallback(callback_);
    if (ret != 0) {
        REQUEST_HILOGE("UnregisterNetConnCallback failed, ret = %{public}d", ret);
    }
}

} // namespace OHOS::Request